tools/autograd/templates/Functions.cpp

```diff
@@ -1970,68 +1970,125 @@ Tensor symeig_backward(const std::vector<torch::autograd::Variable> &grads, cons
+  // above reference.
 
   // Compute R (R')^{T}
-  Tensor R_term;
-  if (grad_R.defined()) {
-    R_term = at::matmul(R, grad_R.transpose(-2, -1));
-  } else {
+    Tensor R_term;
+    if (grad_R.defined()) {
+      R_term = at::matmul(R, grad_R.transpose(-2, -1));
+    } else {
     // R is ... x N x N, grad_R is ... x N x N and grad_R.T is ... x N x N
-    R_term = at::zeros_like(R, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
-  }
+      R_term = at::zeros_like(R, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
+    }
 
-  // Compute Q^{T} Q'
-  Tensor Q_term;
-  if (grad_Q.defined()) {
-    Q_term = at::matmul(Q.transpose(-2, -1), grad_Q);
-  } else {
-    // Q is ... x M x N, Q.T is ... x N x M and grad_Q is ... x M x N
-    Q_term = at::zeros_like(R, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
-  }
-
-  // We want to compute: (rhs_solve_1 . R^{-T})
-  // Note that (rhs_solve_1 . R^{-T}) = (R^{-1} . rhs_solve_1^{T})^{T}
-  // Since R is upper triangular, we can do this using
-  // triangular_solve(rhs_solve_1^{T}, R)^{T}
-  auto rhs_solve_1 = R_term - R_term.transpose(-2, -1) + Q_term - Q_term.transpose(-2, -1);
-  rhs_solve_1 = at::tril(rhs_solve_1, /*k=*/-1);
-  Tensor solve_soln_1;
-  std::tie(solve_soln_1, std::ignore) = at::triangular_solve(rhs_solve_1.transpose(-2, -1), R,
+    // Compute Q^{T} Q'
+    Tensor Q_term;
+    if (grad_Q.defined()) {
+      Q_term = at::matmul(Q.transpose(-2, -1), grad_Q);
+    } else {
+      // Q is ... x M x N, Q.T is ... x N x M and grad_Q is ... x M x N
+      Q_term = at::zeros_like(R, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
+    }
+
+    // We want to compute: (rhs_solve_1 . R^{-T})
+    // Note that (rhs_solve_1 . R^{-T}) = (R^{-1} . rhs_solve_1^{T})^{T}
+    // Since R is upper triangular, we can do this using
+    // triangular_solve(rhs_solve_1^{T}, R)^{T}
+    auto rhs_solve_1 = R_term - R_term.transpose(-2, -1) + Q_term - Q_term.transpose(-2, -1);
+    rhs_solve_1 = at::tril(rhs_solve_1, /*k=*/-1);
+    Tensor solve_soln_1;
+    std::tie(solve_soln_1, std::ignore) = at::triangular_solve(rhs_solve_1.transpose(-2, -1), R,
                                                              /*upper=*/true, /*transpose=*/false,
                                                              /*unitriangular=*/false);
-  Tensor grad_A;
-  if (grad_R.defined()) {
-    grad_A = at::matmul(Q, solve_soln_1.transpose(-2, -1) + grad_R);
-  } else {
-    grad_A = at::matmul(Q, solve_soln_1.transpose(-2, -1));
-  }
-
-  // Successive computations involve computation of QQ^{T} which is identity when A is square
-  if (self.size(-1) != self.size(-2)) {
-    Tensor rhs_solve_2;
-    // We use the same trick from above for this computation
-    if (grad_Q.defined()) {
-      rhs_solve_2 = grad_Q - at::matmul(Q, Q_term);
+    Tensor grad_A;
+    if (grad_R.defined()) {
+      grad_A = at::matmul(Q, solve_soln_1.transpose(-2, -1) + grad_R);
     } else {
-      rhs_solve_2 = -at::matmul(Q, Q_term);
+      grad_A = at::matmul(Q, solve_soln_1.transpose(-2, -1));
```

Comment:
This part might be a little confusing b/c of the diff these lines are all kept in the lambda. Github is splittting these changes across 3 ***patches*** so they git rendered separately the the browser, if you look above at line 2012 that was removed the `grad_R.defined()` control flow is preserved, also the `grad_Q.defined()` control flow is preserved but b/c of the multiple patches the changed lines are getting "striped" so it doesn't appear that they are the same between old and this PR but these lines ***are*** preserved.